src/librbd/internal.cc

```diff
@@ -2248,6 +2310,41 @@ int mirror_image_disable_internal(ImageCtx *ictx, bool force,
     return 0;
   }
 
+  int snap_remove(ImageCtx *ictx, const char *snap_name, uint32_t flags)
+  {
+    ldout(ictx->cct, 20) << "snap_remove " << ictx << " " << snap_name << " flags: " << flags << dendl;
+
+    int r = 0;
+
+retry:
+    r = ictx->state->refresh_if_required();
+    if (r < 0)
+      return r;
+
+    C_SaferCond ctx;
+    ictx->operations->snap_remove(snap_name, &ctx);
+
+    r = ctx.wait();
+    if (r < 0 && r != -EBUSY)
+      return r;
+   
+    if (r == -EBUSY && (flags & RBD_SNAP_REMOVE_UNPROTECT)) {
```

Comment:
Why not just test for needs unprotection or needs children flattening first instead of requiring a goto.